doc/helpers.cmake

```diff
@@ -27,25 +35,92 @@ function(ginkgo_doc_pdf name path)
         )
 endfunction()
 
+
 # generates the documentation named <name> with the additional
 # config file <in> in <pdf/html> format
-function(ginkgo_doc_gen name in pdf mainpage)
+function(ginkgo_doc_gen name in pdf mainpage-in)
     set(DIR_SCRIPT "${CMAKE_CURRENT_SOURCE_DIR}/scripts")
-    set(DIR_BASE "${CMAKE_CURRENT_SOURCE_DIR}/..")
+    set(DIR_BASE "${CMAKE_SOURCE_DIR}")
+    set(DOC_BASE "${CMAKE_CURRENT_SOURCE_DIR}")
     set(DIR_OUT "${CMAKE_CURRENT_BINARY_DIR}/${name}")
-    set(MAINPAGE "${CMAKE_CURRENT_SOURCE_DIR}/pages/${mainpage}")
+    set(MAINPAGE "${DIR_OUT}/MAINPAGE-${name}.md")
     set(doxyfile "${CMAKE_CURRENT_BINARY_DIR}/Doxyfile-${name}")
+    set(layout "${CMAKE_CURRENT_SOURCE_DIR}/DoxygenLayout.xml")
+    ginkgo_file_concat("${CMAKE_CURRENT_SOURCE_DIR}/pages"
+      "${mainpage-in}" BASE_DOC.md "${MAINPAGE}"
+      )
+    set(doxygen_base_input
+      "${CMAKE_CURRENT_SOURCE_DIR}/headers/"
+      )
+    list(APPEND doxygen_base_input
+      ${CMAKE_BINARY_DIR}/include/ginkgo/config.hpp
+      ${DIR_BASE}/include
+      ${MAINPAGE}
+      )
+    if(GINKGO_DOC_GENERATE_EXAMPLES)
+      list(APPEND doxygen_base_input
+        ${CMAKE_CURRENT_BINARY_DIR}/examples/examples.hpp
+        )
+    endif()
+    set(doxygen_dev_input
+      "${DIR_BASE}/core"
+      )
+    list(APPEND doxygen_dev_input
+      ${DIR_BASE}/omp
+      ${DIR_BASE}/cuda
+      ${DIR_BASE}/reference
+      )
+    set(doxygen_image_path "${CMAKE_CURRENT_SOURCE_DIR}/images/")
+    file(GLOB doxygen_depend
+      ${CMAKE_CURRENT_SOURCE_DIR}/headers/*.hpp
+      ${CMAKE_SOURCE_DIR}/include/ginkgo/**/*.hpp
```

Comment:
Could you make these uniform, as sometimes they are `CMAKE_SOURCE_DIR` and other times `DIR_BASE`.